### PR TITLE
Add membership-inference defense and trained-attacker evaluation

### LINKED ISSUE #981: Add a membership-inference defense and trained-attacker evaluation for released de-identification models
## Summary
The existing membership-inference probe is a simple record-matching heuristic and does not evaluate whether a released detector model leaks training-set membership through confidence/score distributions, nor does it provide a defense. A trained shadow-model attacker can exploit calibration gaps to infer whether a note was in the training corpus, undermining the on-device privacy story.

## Scope
- [ ] Implement a shadow-model membership-inference attacker that trains on detector score/logit features over in-corpus vs. held-out synthetic notes and reports attacker AUC and advantage.
- [ ] Add a defense: score-distribution smoothing / confidence clipping / temperature recalibration applied at inference, configurable per policy, that reduces in/out distinguishability without dropping recall below floor.
- [ ] Extend the attack report metric schema to include attacker AUC, advantage, and a per-label membership-leakage breakdown.
- [ ] Wire a membership-leakage release gate that fails when attacker advantage exceeds a configured ceiling.
- [ ] Ensure the defense interacts correctly with existing calibration thresholds and does not regress detection recall.
- [ ] Persist only aggregate scores and hashes in reports, never raw note text.

## Acceptance criteria
- [ ] On the synthetic suite, post-defense membership-inference attacker advantage is at or below the configured ceiling while direct-identifier recall stays at or above the recall floor.
- [ ] The attacker measurably beats random (advantage well above 0) on the undefended baseline, proving the evaluation is non-trivial.
- [ ] Residual direct-identifier leakage under the deidentified outputs remains 0 with the defense enabled.
- [ ] The membership-leakage gate fails a leaky configuration and passes a defended one with a reproducible, signed verdict.
- [ ] Reports contain only aggregate metrics and hashes; a scan asserts no raw PHI is persisted.

## Out of scope
- DP training of the underlying models
- Attacks requiring access to real patient training data (synthetic shadow data only)
- Online/streaming defenses

## Files
- openmed/eval/attacks/reid.py
- openmed/eval/calibrate.py
- openmed/eval/release_gates.py
- openmed/core/thresholds.py

---
Task: OM-618  ·  Milestone: v1.8  ·  Priority: P1  ·  Size: L
Depends on: —  ·  Blocks: —
Roadmap: v1.8 hard initiative
Spec: PLANS/V2/EXECUTION/tasks/OM-618.md

### PR DESCRIPTION
**Pull request by maziyarpanahi**
Closes #981.

Task: OM-618.

Implements a deterministic shadow-score membership-leakage evaluation with aggregate-only metrics, adds configurable score clipping, smoothing, and temperature defense in calibration artifacts and calibrated inference, and wires a signed release-gate check that fails leaky attacker advantage while passing defended configurations.

Validation:
- make lint
- make format-check
- .venv/bin/python -m pytest tests/ -q

### CODE REVIEW

**Comment by maziyarpanahi** [comment]
Reviewed against #981 and current `master`.

Validation:
- `gh pr checks 1037` passed (15 checks passed/skipped).

This satisfies the linked issue and is ready to merge.
